src/main/java/build/buildfarm/metrics/cas/CASAccessMetricsRecorder.java

```diff
@@ -0,0 +1,242 @@
+
+    Map<Digest, AtomicInteger> firstIntervalReadCount = new HashMap<>();
+    if (queueLength > 0 && readIntervalCountQueue.size() > queueLength) {
+      firstIntervalReadCount = readIntervalCountQueue.poll();
+    }
+    firstIntervalReadCount.forEach(
+        (k, v) ->
+            effectiveReadCount.compute(
+                k, (digest, readCount) -> readCount == null ? -v.get() : readCount - v.get()));
+
+    try {
+      Map<String, Integer> updatedReadCount = backplane.updateCasReadCount(effectiveReadCount);
+      expireUnreadEntries(updatedReadCount, firstIntervalReadCount);
+    } catch (Exception e) {
+      log.log(Level.WARNING, "Failed to update the cas read count to backplane", e);
+    }
+
+    long timeToUpdate = stopwatch.stop().elapsed().toMillis();
+    log.fine(format("Took %d ms to update read count.", timeToUpdate));
+
+    if (timeToUpdate >= casEntryReadCountUpdateInterval.toMillis()) {
+      log.warning(
+          "Consider increasing the update interval: read count update time exceeds the update interval.");
+    }
+  }
+
+  private void expireUnreadEntries(
+      Map<String, Integer> casReadCount, Map<Digest, AtomicInteger> firstIntervalReadCount) {
+    Set<Digest> digestsToExpire = new HashSet<>();
+    firstIntervalReadCount.forEach(
+        (digest, readCount) -> {
+          Integer totalReadCount = casReadCount.get(digest.getHash());
+          if (totalReadCount != null && totalReadCount == 0) {
+            digestsToExpire.add(digest);
+            // If the readCount is 0 within an interval, it indicates that the entry was written
+            // during
+            // that interval but remained unread.
+            if (readCount.get() == 0) {
+              log.info(
+                  format(
+                      "Digest %s was not read once after being written in last %s duration",
+                      DigestUtil.toString(digest), casEntryReadCountWindow.toString()));
+            }
+          }
+        });
+    if (digestsToExpire.isEmpty()) {
+      return;
+    }
+    try {
+      int removedCount = backplane.removeCasReadCountEntries(digestsToExpire);
+      log.fine(format("Number of cas read count entries removed : %d", removedCount));
+    } catch (Exception e) {
+      log.log(Level.WARNING, "Failed to remove cas read count entries", e);
+    }
+  }
+
+  @VisibleForTesting
+  Deque<Map<Digest, AtomicInteger>> getReadIntervalCountQueue() {
+    return readIntervalCountQueue;
+  }
+
+  private Map<Digest, AtomicInteger> createNewReadCountInterval() {
+    Map<Digest, AtomicInteger> lastIntervalReadCount = currentIntervalReadCount;
+    currentIntervalReadCount = new ConcurrentHashMap<>();
+    intervalNumber++;
+    return lastIntervalReadCount;
+  }
+
+  private static void runWithLock(Lock lock, Runnable task) {
+    lock.lock();
+    try {
+      task.run();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @SneakyThrows
```

Comment:
Do not use this annotation. Change the signature to accept a Supplier<T>.

Reply:
Done